### PR TITLE
libc: Add GCC fno-builtin-malloc flag to common stdlib compilation

### LINKED ISSUE #64941: gcc optimizes calloc to an infinite recursive call when CONFIG_SPEED_OPTIMIZATION=y
gcc is optimizing out calloc to an infinite recursive call
when CONFIG_COMMON_LIB_CALLOC=y and CONFIG_SPEED_OPTIMIZATIONS=y

See godbolt example (credit @gramsay0 ):
https://godbolt.org/z/s4E9csMa4

@gramsay0 has the following explanation for infinite call:
"I think GCC version 12 -O2 finds and replaces certain code with C std library code.
i.e. a for loop setting values to zero will be replaced with memset, or in this case a call to malloc + memset is replaced by calloc.
The problem being the call to malloc + memset is calloc... so it starts recursing.
I think this would only effect C libraries with CONFIG_COMMON_LIBC_CALLOC=y and CONFIG_COMPILER_FREESTANDING=n (i.e. picolibc)."

**Environment:**
 - OS: Linux
 -  Zephyr SDK (16.1)

To reproduce: 
QEMU board (CONFIG_COMMON_LIBC_CALLOC=y and  CONFIG_SPEED_OPTIMIZATIONS=y) with the following code:
printk("%p\n", calloc(1, 1));

### PR DESCRIPTION
**Pull request by gramsay0**
This prevents the compiler from optimizing calloc into an infinite recursive call.

For example a call to malloc + memset zero at GCC -O2 will be replaced by a call to calloc.
This causes infinite recursion if the function being implemented *is* calloc.

fno-builtin-malloc forces the compiler to avoid this optimization.

This can be tested by adding this patch to the hello_world sample:
```
diff --git a/samples/hello_world/src/main.c b/samples/hello_world/src/main.c
index 2758d75d3f..5841ae99e4 100644
--- a/samples/hello_world/src/main.c
+++ b/samples/hello_world/src/main.c
@@ -5,9 +5,12 @@
  */
 
 #include <stdio.h>
+#include <stdlib.h>
 
 int main(void)
 {
        printf("Hello World! %s\n", CONFIG_BOARD);
+       printf("%p\n", calloc(1, 1));
+       printf("done\n");
        return 0;
 }
```
and building on a QEMU board with optimization -O2:
```
west build -b qemu_cortex_a53 samples/hello_world -t run  -- -DCONFIG_SPEED_OPTIMIZATIONS=y
```

Output before fix:
```
*** Booting Zephyr OS build zephyr-v3.5.0-1326-g4f7f12bdbf69 ***
Hello World! qemu_cortex_a53
```
Output after fix:
```
*** Booting Zephyr OS build zephyr-v3.5.0-1326-g4f7f12bdbf69 ***
Hello World! qemu_cortex_a53
0x4005e890
done
```

Fixes: #64941

### CODE REVIEW

File: lib/libc/common/CMakeLists.txt [hunk 1]
```diff
@@ -6,3 +6,6 @@ zephyr_library_sources_ifdef(CONFIG_COMMON_LIBC_ABORT source/stdlib/abort.c)
 zephyr_library_sources_ifdef(CONFIG_COMMON_LIBC_TIME source/time/time.c)
 zephyr_library_sources_ifdef(CONFIG_COMMON_LIBC_MALLOC source/stdlib/malloc.c)
 zephyr_library_sources_ifdef(CONFIG_COMMON_LIBC_STRNLEN source/string/strnlen.c)
+
+# Prevent compiler from optimizing stdlib functions into infinite recursive calls
+zephyr_library_cc_option(-fno-builtin)
```
Comment (keith-packard) [inline lib/libc/common/CMakeLists.txt:11]: I build code like this with only the necessary builtins disabled. That allows the compiler to continue to perform optimizations and checking for other functions. In this case, adding `-fno-builtin-malloc -fno-builtin-free` would be a more specific change and will solve the problem.

**Review by keith-packard — CHANGES_REQUESTED** [review]

File: lib/libc/common/CMakeLists.txt (hunk 1 above)
Comment (gramsay0) [reply lib/libc/common/CMakeLists.txt:11]: I agree that would fix this issue. My concern would be that after the next GCC release, this starts occurring with `strlen` and the other functions in here. There is not much code in this common libc library to optimize, and every function is probably somewhat susceptible to this issue

Comment (keith-packard) [reply lib/libc/common/CMakeLists.txt:11]: yeah, it's a balance -- using the big hammer (`-fno-builtin`) gets rid of all checking and optimizations for all functions the compiler knows about, which makes code worse (both in performance and in correctness). Using specific options means you might miss one and cause recursion issues like this. On balance, for picolibc, I decided that I was willing to accept the latter risk as it's quite easy to test for while the former is essentially un-testable. Of course, that relies on having basic tests for each of these functions to make sure they don't recurse.

Comment (gramsay0) [reply lib/libc/common/CMakeLists.txt]: Fair enough, I have modified the PR to use `-fno-builtin-malloc`

Comment (keith-packard) [reply lib/libc/common/CMakeLists.txt]: Looks like we should add a bug to add tests for each one of these functions someday?

**Review by keith-packard — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]

File: lib/libc/common/CMakeLists.txt [hunk 2]
```diff
@@ -6,3 +6,6 @@ zephyr_library_sources_ifdef(CONFIG_COMMON_LIBC_ABORT source/stdlib/abort.c)
 zephyr_library_sources_ifdef(CONFIG_COMMON_LIBC_TIME source/time/time.c)
 zephyr_library_sources_ifdef(CONFIG_COMMON_LIBC_MALLOC source/stdlib/malloc.c)
 zephyr_library_sources_ifdef(CONFIG_COMMON_LIBC_STRNLEN source/string/strnlen.c)
+
+# Prevent compiler from optimizing calloc into an infinite recursive call
+zephyr_library_cc_option(-fno-builtin-malloc)
```
Comment (stephanosio) [inline lib/libc/common/CMakeLists.txt:11]: This needs to be properly abstracted for multiple toolchain support.

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: lib/libc/common/CMakeLists.txt (hunk 2 above)
Comment (gramsay0) [reply lib/libc/common/CMakeLists.txt]: @stephanosio I gave this a go, is it what you were wanting?
I am not familiar with some of these compilers, so I may have made a mistake. Synopsys seems to require an account to look at arcmwdt supported flags (https://www.synopsys.com/dw/doc.php/arc_tools/dw_arc_metaware/hw_id_core_versions.pdf)

Comment (stephanosio) [reply lib/libc/common/CMakeLists.txt]: > I gave this a go, is it what you were wanting?

Thanks! That is exactly what I meant.

> I am not familiar with some of these compilers, so I may have made a mistake. Synopsys seems to require an account to look at arcmwdt supported flags

For now, only adding the support for the affected compiler(s) should be fine. The support for other proprietary compilers can be added later by the relevant maintainers (assuming they also require this flag). cc @abrodkin @evgeniy-paltsev

**Review by stephanosio — APPROVED** [review]

**Review by aescolar — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]

**Pull request merged**